File: LinkedLists/linked_lists.py
# linked_lists.py
"""Volume 2: Linked Lists.
Caelan Osman
Math 345 Sec 3
October 3rd, 2020
"""


# Problem 1
class Node:
    """A basic node class for storing data."""
    def __init__(self, data):
        """Store the data in the value attribute.
        Raises:
            TypeError: if data is not of type int, float, or str.
        """
        if type(data) != int and type(data) != str and type(data) != float:
            raise TypeError('data is not of type "str", "int", or "float"')
        else:
            self.value = data

class LinkedListNode(Node):
    """A node class for doubly linked lists. Inherits from the Node class.
    Contains references to the next and previous nodes in the linked list.
    """
    def __init__(self, data):
        """Store the data in the value attribute and initialize
        attributes for the next and previous nodes in the list.
        """
        Node.__init__(self, data) # Use inheritance to set self.value.
        self.next = None # Reference to the next node.
        self.prev = None # Reference to the previous node.


# Problems 2-5
class LinkedList:
    """Doubly linked list data structure class.

    Attributes:
        head (LinkedListNode): the first node in the list.
        tail (LinkedListNode): the last node in the list.
    """
    def __init__(self):
        """Initialize the head and tail attributes by setting
        them to None, since the list is empty initially.
        """
        self.head = None
        self.tail = None
        self.length = 0

    def append(self, data):
        """Append a new node containing the data to the end of the list."""
        new_node = LinkedListNode(data) # Create a new node to store the input data.
        self.length += 1
        if self.head is None: # If the list is empty, assign the head and tail attributes to
            self.head = new_node
            self.tail = new_node
        else:
            # If the list is not empty, place new_node after the tail.
            self.tail.next = new_node # tail --> new_node
            new_node.prev = self.tail # tail <- new_node
            self.tail = new_node #reassign the tail

    # Problem 2
    def find(self, data):
        """Return the first node in the list containing the data.

        Raises:
            ValueError: if the list does not contain the data.
        """
        #edge case if the list is empty head is None then the list is empty and the linked list cannot contain the node.
        if self.head == None:
            raise ValueError('The node is not contained in the Linked list')

        #set intial conditions
        foundnode = None
        n = self.head
        while(foundnode is None): #repeat until we find node
            if data == n.value: #if data is equal to the node value then return that node
                return n
            elif n == self.tail: #if we are at the last node, exit the loop
                foundnode = False
            else:
                n = n.next #otherwise check next node

        #if the node is not found then this error is raised
        raise ValueError('Data is not contained in Linked List')

    # Problem 2
    def get(self, i):
        #if the index is larger than the size of the list through an out of range error
        if i < 0 or i >= self.length:
            raise IndexError('"Index" out of range')

        #otherwise loop through and get the node
        else:
            n = self.head
            if i == 0:
                return n
            n = n.next
            for j in range(0, i):
                if j == i-1:
                    return n
                else:
                    n = n.next

    # Problem 3
    def __len__(self):
        """Returns the number of nodes in the list.
        """
        return self.length

    # Problem 3
    def __str__(self):
        """This function returns a string representation of the liked list
           which is the same as the standard python list
        """

        #starting node and initial string representation
        n = self.head
        string_rep = '['

        if self.length == 0: #returns an "empty list" if the length is zero
            return string_rep + ']'
        elif self.length == 1: #if the lenght of the list is one
            if type(n.value) == str:
                 string_rep += repr(n.value) + ']'
                 return string_rep
            else:
               string_rep += str(n.value) + ']'
               return string_rep

        #if the length of the list is greater than one.
        for i in range(0, self.length): #loop through nodes
            if type(n.value) == str: #if the node type is string
                if i == 0: #deals with comma and bracket placement
                    string_rep += repr(n.value)
                elif i == self.length - 1:
                    string_rep += ', ' + repr(n.value) +']'
                else:
                    string_rep += ', ' + repr(n.value)
            else: #same as above but for numbers
                if i == 0:
                    string_rep += str(n.value)
                elif i == self.length - 1:
                    string_rep += ', ' + str(n.value) +']'
                else:
                    string_rep += ', ' + str(n.value)
            n = n.next #set our node to the next node in the linked list
        return string_rep

    # Problem 4
    def remove(self, data):
        """Removes the first node in the list containing the data.

        Raises:
            ValueError: if the list is empty or does not contain the data.
        """

        #if the list is empty
        if self.length == 0:
            raise ValueError('The list is empty')

        #special case where the list only has one element
        n = self.head
        if self.length == 1 and data == self.head.value:
            self.head = None
            self.tail = None
            self.length = 0
            return
        #special case where the len(llist) > 1 and the data is the first node
        elif data == n.value:
            self.head = n.next
            self.head.prev = None
            self.length -= 1
            return
        #special case where len(llist) > 1 and the data is the final node
        elif data == self.tail.value:
            self.tail = self.tail.prev
            self.tail.next = None
            self.length -= 1
            return
        else:
            n = n.next
            #only need to loop from the 2nd to n-1 nodes because of ^^
            for i in range(1, self.length-1):
                if data == n.value:
                    #next of the previous and the previous of the next
                    n.prev.next = n.next
                    n.next.prev = n.prev
                    self.length -= 1
                    return
                #otherwise loop onto the next
                else:
                    n = n.next

        raise ValueError('Data not found') #finally if the data is not found

    # Problem 5
    def insert(self, index, data):
        """Insert a node containing data into the list immediately before the
        node at the index-th location.

        Raises:
            IndexError: if index is negative or strictly greater than the
                current number of nodes.
        """
        #throws an out of range error if the index doesn't make sense
        if index < 0 or index > self.length:
            raise IndexError('"Index" Out of range')
        #special case where we are inserting and resetting the head and the list is not empty
        if index == 0 and self.length > 0:
            new = LinkedListNode(data) #construct new node
            new.next = self.head #set the pointer to the current head
            new.next.prev = new  #set the pointer of the next back to the new
            self.head = new #set the head as the new
            self.length += 1 #update length
            return
        #Special case where we are inserting and resetting the head and the list is empty
        elif index == 0 and self.length == 0:
            new = LinkedListNode(data)
            self.head = new
            self.tail = new
            self.length = 1
            return
        #if the insert is the length of the list just append
        elif index == self.length:
            self.append(data)
            return
        #do the same as in the first elif statement except at an index > 1
        else:
            n = self.head
            for j in range(0,index-1):
                n = n.next
            new = LinkedListNode(data)
            new.prev = n
            new.next = n.next
            n.next = new
            new.next.prev = new
            self.length += 1
            return

class Deque(LinkedList):
    '''Deque data structure class:

       Attributes:
           Head Node
           Tail Node
           length of deque
    '''

    #use LinkedList class constructor
    def __init__(self):
        LinkedList.__init__(self)

    def pop(self): #function to remove last node
        if self.head == None: #edge case if deque is empty
            raise ValueError('The deque is empty, cannot pop.')
        elif self.length == 1: #edge case if deque only has one node
            node_value = self.tail.value
            self.tail = None
            self.head = None
            self.length = 0
            return node_value
        else: #otherwise reset tail value
            node_value = self.tail.value
            self.tail = self.tail.prev
            self.tail.next = None
            self.length -= 1
            return node_value

    def popleft(self): #function to remove first node
        if self.head == None: #edge case if our deque is emtpy
            raise ValueError('The deque is empty, cannot popleft.')
        elif self.length == 1: #edge case if our deque only has one element
            node_value = self.head.value
            self.tail = None
            self.head = None
            self.length = 0
            return node_value
        #use inhertiance to remove node
        else:
            node_value = self.head.value
            LinkedList.remove(self, node_value)
            return node_value
    #if we want to add nodes to the beginning of our deque we just use
    #inheritence on the insert function with index 0.
    def appendleft(self, data):
        LinkedList.insert(self, 0, data)

    #these just remove functionality that we don't want to be able
    #to be accessed because deques are restricted access
    def remove(*args, **kwargs):
        raise NotImplementedError('Use pop() or popleft() for removal')

    def insert(*args, **kwargs):
        raise NotImplementedError('Use append() or appendleft()')

# Problem 7
def prob7(infile, outfile):
    """Reverse the contents of a file by line and write the results to
    another file.

    Parameters:
        infile (str): the file to read from.
        outfile (str): the file to write to.
    """
    filedeque = Deque()
    with open(infile, 'r') as ip_file:
        contents = ip_file.read()
        splitline = contents.strip().split('\n')
        for i in range(len(splitline)):
            filedeque.append(str(splitline[i]))

    print(filedeque)
    final_string = ''
    with open(outfile, 'w') as op_file:
        for i in range(filedeque.length):
            final_string += str(filedeque.pop()) + '\n'
            #op_file.write(str(filedeque.pop()) + '\n')

    return final_string


    '''
    final_string = ''
    filedeque = Deque() #create our deque
    with open(infile, 'r') as ip: #open our file
        lines = ip.read().strip().split('\n') #add all lines to a list
    for i, line in enumerate(lines): #populate our deque
        #if we are at the last element make sure to add another newline character
        if i == len(lines) - 1:
            filedeque.appendleft('\n' + line + '\n')
        #make sure to not add an additional newline
        elif i == 0:
            filedeque.appendleft(line)
        else:
            filedeque.appendleft(line + '\n')
    #assemble string
    for _ in range(filedeque.length):
        final_string += filedeque.popleft()
    #write to our outfile
    with open(outfile, 'w') as of:
        of.write(final_string)

    '''

if __name__ == '__main__':

    backwards_file_string = '\noutlooks\neuthanasia\noutlier\nyours\nglistened\nsociolinguistics\nfixations\ndoubts\nencyclopaedic\nleer'
    deque = Deque()
    make_lines = backwards_file_string.split('\n')
    for line in make_lines:
        deque.append(line)
    with open('test.txt', 'w') as key:
        for _ in range(deque.length):
            key.write(deque.pop() + '\n')

    string = prob7('test.txt', 'test_output.txt')
    print(repr(string))

File: Julia/LinkedLists/linked_lists.jl
# linked_lists.jl
module LinkedLists

import Base.repr, Base.append!, Base.==, Base.insert!

export LinkedListNode, LinkedList, Deque, append!, find, get, length, repr, remove!, pop!, popleft!, insert!, appendleft!, ==

# abstract type Node end
# mutable struct LinkedListNode <: Node
mutable struct LinkedListNode
    data::Union{Nothing, <:Number, String, Vector{<:Any}}
    next::Union{Nothing,LinkedListNode}
    prev::Union{Nothing,LinkedListNode}
end

abstract type CustomDataStructure end

mutable struct LinkedList <: CustomDataStructure

    head::Union{Nothing,LinkedListNode}
    tail::Union{Nothing,LinkedListNode}
    nodeCount::Int64

    function LinkedList(head,tail,nodeCount)
        if isnothing(head)
            nodeCount = 0
        else
            nodeCount = 1
        end
        return new(head,tail,nodeCount)
    end
end

mutable struct Deque <: CustomDataStructure

    head::Union{Nothing,LinkedListNode}
    tail::Union{Nothing,LinkedListNode}
    nodeCount::Int64

    function Deque(head,tail,nodeCount)
        if isnothing(head) && isnothing(tail)
            nodeCount = 0
        else
            nodeCount = 1
        end
        return new(head,tail,nodeCount)
    end
end

function append!(DS::CustomDataStructure, data)

    # initialize the new node using the data
    newNode = LinkedListNode(data,nothing,nothing)

    if isequal(DS.head,nothing)
        # check to see if the linked list is empty
        # if it is assing the new node to be the head and tai
        DS.head = newNode
        DS.tail = newNode
        DS.nodeCount += 1
    else
        # if the linked list is not empty then the new node goes at
        # the end of the list
        DS.tail.next = newNode # connect the new node to the end of the list
        newNode.prev = DS.tail # make it so the previous of the new node is the old tail
        DS.tail = newNode # reassign the tail to be the new node
        DS.nodeCount += 1
    end

    return
end

function find(DS::CustomDataStructure, data::Union{Nothing, <:Number, String, Vector{<:Any}})

    # check to see if the DS is empty
    if isequal(DS.head,nothing)
        throw(KeyError(data))
    end
    # now we search through the list
    # set intiial node
    currentNode = DS.head
    search=true
    while search
        # checkt to see if the data inside the DSNode is equivalent
        # to the given data
        if currentNode.data == data
            search = false
        else
            # get the next node in the Linked List
            currentNode = currentNode.next
            if isequal(currentNode, nothing)
                search = false
                throw(KeyError(data))
            end
        end
    end
    return currentNode
end

function get(DS::CustomDataStructure, index::Int64)

    if index < 1
        throw(KeyError(index))
    end
    if index > DS.nodeCount
        throw(KeyError(index))
    end

    currentNode = DS.head
    for i=2:index
        currentNode = currentNode.next
    end

    return currentNode
end


length(DS::CustomDataStructure) = DS.nodeCount

function ==(CDS1::CustomDataStructure, CDS2::CustomDataStructure)


    if !isequal(typeof(CDS1),typeof(CDS2))
        return false
    elseif CDS1.nodeCount ≠ CDS2.nodeCount
        return false
    else
        currentNode1 = CDS1.head
        currentNode2 = CDS2.head
        for i=2:CDS1.nodeCount
            if currentNode1.data ≠ currentNode2.data
                return false
            end
                currentNode1 = currentNode1.next
                currentNode2 = currentNode2.next

        end
    end
    return true
end

function repr(DS::CustomDataStructure)

    currentNode = DS.head
    stringRep = "["

    if DS.nodeCount == 0
        stringRep *= "]"
        return Base.repr(stringRep);
    elseif DS.nodeCount == 1
        if typeof(currentNode.data) == String
            stringRep *= Base.repr(currentNode.data) * "]";
        else
            stringRep *= string(currentNode.data) * "]";
        end
        return stringRep
    end

    for i=1:DS.nodeCount
        if typeof(currentNode.data) == String
            if i == 1
                stringRep *= Base.repr(currentNode.data);
            elseif i == DS.nodeCount
                stringRep *= ", " * Base.repr(currentNode.data) * "]";
            else
                stringRep *= ", " * Base.repr(currentNode.data);
            end
        else
            if i == 1
                string_rep *= string(currentNode.data);
            elseif i == DS.nodeCount
                stringRep *= ", " * string(currentNode.data) * "]"''
            else
                stringRep *= ", " * string(currentNode.data);
            end
        end
        currentNode = currentNode.next
    end

    return stringRep
end


function remove!(DS::CustomDataStructure, data::Union{Nothing,<:Number,String, Vector{<:Any}})

    if typeof(DS) == Deque
        throw(ArgumentError("Use pop!() or popleft!() for removal."))
    end

    if DS.nodeCount == 0
        throw(KeyError(DS.nodeCount))
    end

    currentNode = DS.head

    # special case where the list only has one element
    if DS.nodeCount == 1  && data == currentNode.data
        DS.head = nothing
        DS.tail = nothing
        DS.nodeCount = 0
        return

    # special case where the linked list has more than one element and the data to remove is the first node.
    elseif data == currentNode.data
        DS.head = currentNode.next
        DS.head.prev = nothing
        DS.nodeCount -= 1
        return

    else
        for i=2:DS.nodeCount
            currentNode = currentNode.next
            # special case where data to remove is the final node
            if data == currentNode.data
                if i == DS.nodeCount
                    DS.tail = DS.tail.prev
                    DS.tail.next = nothing
                    DS.nodeCount -= 1
                    return
                else
                    currentNode.prev.next = currentNode.next
                    currentNode.next.prev = currentNode.prev
                    DS.nodeCount -= 1
                    return
                end
            end
        end
    end
    throw(KeyError(data))
end

function pop!(D::Deque)
    # function remove last node
    if D.nodeCount == 0 # edge case if deque is empty
        throw(ArgumentError(D, "Index out of range"))
    elseif D.nodeCount == 1 #edge case if deque only has one node
        nodeData = D.tail.data
        D.tail = nothing
        D.head= nothing
        D.nodeCount -= 1
    else # otherwise reset the tail value
        nodeData = D.tail.data
        D.tail = D.tail.prev
        D.tail.next = nothing
        D.nodeCount -= 1
    end
    return nodeData
end

function popleft!(D::Deque)
    # function to remove first node
    if D.nodeCount == 0 # edge case if our deque is empty
        throw(ArgumentError(D, "Index out of range"))
    elseif D.nodeCount == 1 # edge case if our deque only has one elemnt
        nodeData = D.head.data
        D.tail = nothing
        D.head = nothing
        D.nodeCount -= 1
    else # otherwise we can just use remove.
        nodeData = D.head.data
        remove!(D, nodeData)
    end

    return nodeData
end


function insert!(DS::CustomDataStructure, index::Int64, data)
    if typeof(DS) == Deque
        throw(ArgumentError(DS, "Use append!() or appendleft!() for insertion."))
    end

    if index < 1 || index > DS.nodeCount + 1
        throw(KeyError(index, "index out of range."))

    end

    # create the new node
    newNode = LinkedListNode(data, nothing,nothing)
    # special case where we are inserting and resetting the head and the DS is not empty.
    if index == 1 && DS.nodeCount > 0
        newNode.next = DS.head
        newNode.next.prev = newNode
        DS.head = newNode
        DS.nodeCount += 1

    # special case where we are inserting and resetting the head and the DS is empty.
    elseif index == 1 && DS.nodeCount == 0
        DS.head = newNode
        DS.tail = newNode
        DS.nodeCount += 1

   # if the insert is just adding on a new node then we can just append
   elseif index == DS.nodeCount + 1
        append!(DS, data)
   # the final case where the insertion point is in the middle of the Linked List
   else
        currentNode = DS.head
        for i=1:index-1
            currentNode = currentNode.next
        end
        newNode.prev = currentNode
        newNode.next = currentNode.next
        currentNode.next = newNode
        newNode.next.prev = newNode
        DS.nodeCount += 1
    end
    return
end

function appendleft!(D::Deque, data)
    insert!(D,0,data)
    return
end
end
